File: plugin/src/main/groovy/com/sensorsdata/analytics/android/plugin/js/AddJSAnnotationVisitor.java
package com.sensorsdata.analytics.android.plugin.js;


import com.sensorsdata.analytics.android.plugin.SensorsAnalyticsTransformHelper;
import com.sensorsdata.analytics.android.plugin.SensorsAnalyticsUtil;

import org.objectweb.asm.AnnotationVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.commons.AdviceAdapter;

public class AddJSAnnotationVisitor extends AdviceAdapter {
    private final SensorsAnalyticsTransformHelper mTransformHelper;
    boolean shouldAddUCJS = false;
    boolean shouldAddXWalkJS = false;

    public AddJSAnnotationVisitor(MethodVisitor mv, int access, String name, String desc, SensorsAnalyticsTransformHelper transformHelper) {
        super(SensorsAnalyticsUtil.ASM_VERSION, mv, access, name, desc);
        this.mTransformHelper = transformHelper;
    }

    @Override
    public AnnotationVisitor visitAnnotation(String s, boolean b) {
        if (s.equals("Landroid/webkit/JavascriptInterface;")) {
            shouldAddUCJS = mTransformHelper.getExtension().addUCJavaScriptInterface;
            shouldAddXWalkJS = mTransformHelper.getExtension().addXWalkJavaScriptInterface;
        }

        return super.visitAnnotation(s, b);
    }

    @Override
    public void visitEnd() {
        super.visitEnd();
        if (shouldAddUCJS) {
            visitAnnotation("Lcom/uc/webview/export/JavascriptInterface;", true);
        }
        if (shouldAddXWalkJS) {
            visitAnnotation("Lorg/xwalk/core/JavascriptInterface;", true);
        }
    }
}

File: sa-gradle-plugin/src/main/kotlin/com/sensorsdata/analytics/android/plugin/js/AddJSAnnotationVisitor.kt
package com.sensorsdata.analytics.android.plugin.js

import com.sensorsdata.analytics.android.plugin.manager.SAPluginManager
import org.objectweb.asm.AnnotationVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.commons.AdviceAdapter

class AddJSAnnotationVisitor(mv: MethodVisitor?, access: Int, name: String?, desc: String?, private val pluginManager: SAPluginManager) :
    AdviceAdapter(pluginManager.getASMVersion(), mv, access, name, desc) {
    var shouldAddUCJS = false
    var shouldAddXWalkJS = false
    override fun visitAnnotation(s: String, b: Boolean): AnnotationVisitor {
        if (s == "Landroid/webkit/JavascriptInterface;") {
            shouldAddUCJS = pluginManager.extension.addUCJavaScriptInterface
            shouldAddXWalkJS = pluginManager.extension.addXWalkJavaScriptInterface
        }
        return super.visitAnnotation(s, b)
    }

    override fun visitEnd() {
        super.visitEnd()
        if (shouldAddUCJS) {
            visitAnnotation("Lcom/uc/webview/export/JavascriptInterface;", true)
        }
        if (shouldAddXWalkJS) {
            visitAnnotation("Lorg/xwalk/core/JavascriptInterface;", true)
        }
    }
}